### PR TITLE
Feature/properties save themselves

### PR DESCRIPTION
**Pull request by JohnGilb**

### CODE REVIEW

File: src/xbot/common/properties/DatabaseStorageBase.java [hunk 1]
```diff
@@ -13,25 +13,140 @@
...
+    }
+    
+    public Boolean getBoolean (String key){
+        String value = loadProperty(key);
+        return parseBoolean(value);
+    }
+    
+    private Boolean parseBoolean(String value)
+    {
+        if (value.toLowerCase().equals("true"))
```
Comment (aschokking) [inline src/xbot/common/properties/DatabaseStorageBase.java]: Doesn't Java have a thing for this already maybe?

File: src/xbot/common/properties/DatabaseStorageBase.java [hunk 2]
```diff
@@ -13,25 +13,140 @@
...
+        return value;
+    }
+    
+    public void clear(){
+        obliterateStorage();
+    }
+    
+    private void saveProperty(String type, String name, String value) {
+       
+        String payload = "UPDATE PROPERTIES SET TYPE='" + name + "', VALUE='" + type + "' WHERE NAME = '" + value + "'";
```
Comment (aschokking) [inline src/xbot/common/properties/DatabaseStorageBase.java]: A little sanitization here would make a lot of sense, if nothing else make sure no ' ?

File: src/xbot/common/properties/DatabaseStorageBase.java [hunk 3]
```diff
@@ -13,25 +13,140 @@
...
+            log.warn(e.toString());
+            e.printStackTrace();
+        }
+    }
+    
+    private String loadProperty(String name)
+    {
+        try {
+            Statement sta = conn.createStatement();
+            String payload = "SELECT * FROM PROPERTIES WHERE NAME = '" + name + "'";
```
Comment (aschokking) [inline src/xbot/common/properties/DatabaseStorageBase.java]: I can't remember, is there some protection to make sure names are unique

File: src/xbot/common/properties/DatabaseStorageBase.java (hunk 1 above)
Comment (JohnGilb) [inline src/xbot/common/properties/DatabaseStorageBase.java]: I'll add a comment.

File: src/xbot/common/properties/DatabaseStorageBase.java [hunk 4]
```diff
@@ -52,11 +169,27 @@ private boolean propertiesTableExists(Connection conn) {
         return false;
     }
 
+    private boolean createPropertiesTable() {
+        
+        Statement sta;
+        try {
+            sta = conn.createStatement();
+            String payload = "CREATE TABLE PROPERTIES (Name VARCHAR(100), Type VARCHAR(20), Value VARCHAR(50))";
```
Comment (aschokking) [inline src/xbot/common/properties/DatabaseStorageBase.java]: yeah no uniqueness constraint?

File: src/xbot/common/properties/DatabaseStorageBase.java (hunk 2 above)
Comment (JohnGilb) [inline src/xbot/common/properties/DatabaseStorageBase.java]: I found a cool thing here, I'll take advantage of it: http://stackoverflow.com/questions/1812891/java-escape-string-to-prevent-sql-injection

File: src/xbot/common/properties/DatabaseStorageBase.java (hunk 3 above)
Comment (JohnGilb) [inline src/xbot/common/properties/DatabaseStorageBase.java]: I'll see if, during table creation, I can easily define the "names" column as having to be unique.

File: src/xbot/common/properties/DatabaseStorageBase.java (hunk 4 above)
Comment (JohnGilb) [inline src/xbot/common/properties/DatabaseStorageBase.java]: See above

File: src/xbot/common/properties/DatabaseStorageBase.java [hunk 5]
```diff
@@ -13,25 +13,144 @@
...
         dbUrl = String.format(dbUrlPreFormat, databaseDirectory);
+        
+        try {
+            conn = DriverManager.getConnection(dbUrl);
+        } catch (SQLException e) {
+            log.error("Ran into a SQL problem - could not open a connection to the database! No properties will be loaded or persisted!!");
+            log.error(e.toString());
+            e.printStackTrace();
+        } catch (Exception e) {
+            log.error("Ran into a general expection, so could not open a connection to the database! No properties will be loaded or persisted!!");
```
Comment (WasabiFan) [inline src/xbot/common/properties/DatabaseStorageBase.java]: "expection" -> "exception"

File: src/xbot/common/properties/DatabaseStorageBase.java [hunk 6]
```diff
@@ -13,25 +13,144 @@
...
+    }
+    
+    public Double getDouble(String key){
+        String value = loadProperty(key);
+        try {
+            return Double.parseDouble(value);
+        }
+        catch (NumberFormatException e)
+        {
+            return null;
```
Comment (WasabiFan) [inline src/xbot/common/properties/DatabaseStorageBase.java]: Do we want to just return null here? If There's a string key and something tries to parse it as a double, can't we be pretty sure that something went horribly wrong?

File: src/xbot/common/properties/DatabaseStorageBase.java [hunk 7]
```diff
@@ -13,25 +13,144 @@
...
+                rs.close();
+                return value;
+            }
+            rs.close();
+        } catch (SQLException e) {
+            log.warn("Unable to load property " + name + "!");
+            log.warn(e.toString());
+            e.printStackTrace();
+        }
+        return "";
```
Comment (WasabiFan) [inline src/xbot/common/properties/DatabaseStorageBase.java]: Empty string?

File: src/xbot/common/properties/DatabaseStorageBase.java (hunk 5 above)
Comment (JohnGilb) [inline src/xbot/common/properties/DatabaseStorageBase.java]: roger

**Comment by aschokking** [comment]
lgtm

File: src/xbot/common/properties/DatabaseStorageBase.java (hunk 4 above)
Comment (JohnGilb) [inline src/xbot/common/properties/DatabaseStorageBase.java]: Added

**Pull request merged**